docs/source/learn/trace-vs-execution-time.md

```diff
@@ -0,0 +1,175 @@
+builds this graph on the host (CPU).
+
+1. Device Time: The period during which the XLA device (e.g., TPU) actually
+performs the computations. This can be further broken down:
+
+   - Compilation Time: The time XLA takes to compile the traced graph into
+   optimized device code. This is most significant on the first execution of a
+   new graph or if the graph changes.
+
+   - Execution Time: The time the XLA device spends running the compiled code.
+
+**If you measure the duration of code that only performs XLA operations without an
+explicit instruction to wait for the device, you are primarily measuring this
+tracing time plus Python overhead. This is NOT an accurate measure of your
+model's performance on the XLA device.**
+
+## Tracing: Defining the Computation (Host-Side Activity)
+
+When you write PyTorch code using XLA tensors (e.g., tensors on a TPU),
+PyTorch/XLA doesn't execute each operation on the device right away. It traces
+these operations, adding them to an internal computation graph. If you measure
+the duration of code that only performs XLA operations without an explicit
+instruction to wait for the device, you are primarily measuring this tracing
+time plus Python overhead.
+
+### Illustrating a Common Pitfall: Measuring Only Tracing Time
+
+Consider the following conceptual code:
+
+```Python
+# Assume 'a' and 'b' are XLA tensors
+# start_time = time.perf_counter()
+
+# This operation is recorded in PyTorch/XLA's graph
+result = torch.matmul(a, b)
+
+#!!! INCORRECT PROFILING: compilation and execution are deferred !!!
+# end_time = time.perf_counter()
+# elapsed_time = end_time - start_time
+```
+
+The `elapsed_time` here would predominantly reflect how long it took PyTorch/XLA
+to trace the matmul operation. The actual matrix multiplication on the XLA
+device is not started.
+
+## Device Time: Compilation and Execution on the Accelerator
+
+To correctly profile the performance of your code on the XLA device, you must
+ensure that your timing captures the phase where the device actually does the
+work. This involves:
+
+1. Ensure the traced computational graph is sent to the device for execution.
+
+1. Make sure the Python script wait until the XLA device has completed all its
+assigned computations before taking the final timestamp.
+
+### Triggering Execution and Ensuring Completion
+
+Several mechanisms trigger graph execution and/or ensure completion:
+
+1. `torch_xla.sync(wait=True)`: This is the most direct method for benchmarking.
+It ensures all pending XLA operations are launched and, crucially, blocks the
+Python script until the device finishes.
+
+1. `Data Access/Transfer`: Operations like `tensor.cpu()`, `tensor.item()`, or
+printing an XLA tensor require the actual data. To provide it, PyTorch/XLA must
+execute the graph that produces the tensor and wait for its completion.
+
+1. `torch_xla.core.xla_model.mark_step()`: Commonly used in loops, this signals
```

Comment:
`mark_step()` is synonymous with `torch_xla.sync`. I think we can merge their descriptions.

Reply:
@tengyifei My motivation behind keeping the subtle difference  `mark_step` and `sync` 
```
mark_step: doesn't inherently block Python until device completion.

VS

sync: blocks the Python script until the device finishes.
```

to illustrate the need for `sync` in the example at https://github.com/pytorch/xla/issues/8858#issue-2933324391. 


wdyt?

Reply:
nvm. I got to know about this change https://github.com/pytorch/xla/pull/9086 recently. With that your suggestion makes sense.